securestring module parameter takes key vault secret reference

miqm opened the pull request:
## Contributing a feature

* [x] I have opened a new issue for the proposal, or commented on an existing one, and ensured that the bicep maintainers are good with the design of the feature being implemented
* [x] I have included "Fixes #{issue_number}" in the PR description, so GitHub can link to the issue and close it when the PR is merged
* [ ] I have appropriate test coverage of my new feature

Fixes #1028

miqm commented:
@majastrz @anthony-c-martin @shenglol @TarunSunkaraneni This is my initial implementation of referencing secrets in module parameters. I've created this PR as a draft, it still lacks tests and probably support for decorators. However I wanted to get your opinion on the approach in introducing a type 'secret'.

My idea for implementation was that the function `getSecret` will return type (`secret`) and the module parameter, if decorated with `{secret: true}` will accept union of plain text and the function output: `string | secret`.

Is this the good approach?

alex-frankel commented:
Thank you for getting this done!!

Did some testing with these files:

`main.bicep` which creates a keyvault and secret:

```bicep
param vaultName string = 'keyVault${uniqueString(resourceGroup().id)}' // must be globally unique
param location string = resourceGroup().location
param sku string = 'Standard'
param tenant string = '72f988bf-86f1-41af-91ab-2d7cd011db47' // replace with your tenantId
param accessPolicies array = [
  {
    tenantId: tenant
    objectId: 'caeebed6-cfa8-45ff-9d8a-03dba4ef9a7d' // replace with your objectId
    permissions: {
      keys: [
        'Get'
        'List'
        'Update'
        'Create'
        'Import'
        'Delete'
        'Recover'
        'Backup'
        'Restore'
      ]
      secrets: [
        'Get'
        'List'
        'Set'
        'Delete'
        'Recover'
        'Backup'
        'Restore'
      ]
      certificates: [
        'Get'
        'List'
        'Update'
        'Create'
        'Import'
        'Delete'
        'Recover'
        'Backup'
        'Restore'
        'ManageContacts'
        'ManageIssuers'
        'GetIssuers'
        'ListIssuers'
        'SetIssuers'
        'DeleteIssuers'
      ]
    }
  }
]

param enabledForDeployment bool = true
param enabledForTemplateDeployment bool = true
param enabledForDiskEncryption bool = true
param enableRbacAuthorization bool = false
param softDeleteRetentionInDays int = 90

param keyName string = 'prodKey'
param secretName string = 'bankAccountPassword'
param secretValue string = '12345'

param networkAcls object = {
  ipRules: []
  virtualNetworkRules: []
}

resource keyvault 'Microsoft.KeyVault/vaults@2019-09-01' = {
  name: vaultName
  location: location
  properties: {
    tenantId: tenant
    sku: {
      family: 'A'
      name: sku
    }
    accessPolicies: accessPolicies
    enabledForDeployment: enabledForDeployment
    enabledForDiskEncryption: enabledForDiskEncryption
    enabledForTemplateDeployment: enabledForTemplateDeployment
    softDeleteRetentionInDays: softDeleteRetentionInDays
    enableRbacAuthorization: enableRbacAuthorization
    networkAcls: networkAcls
  }
}
// create secret
resource secret 'Microsoft.KeyVault/vaults/secrets@2018-02-14' = {
  name: '${keyvault.name}/${secretName}'
  properties: {
    value: secretValue
  }
}

module secretDeploy 'module-with-secret.bicep' = {
  name: 'foo'
  params: {
    secret: keyvault.getSecret(secretName) // would also like to provide the secret resource as a valid type
  }
}
```

And a dummy module `module-with-secret.bicep` to pass the secret to:

```bicep
// @secure() did not work
param secret string {
  secure: true
}

output badPractice string = secret
```

I hit a bug if I use a decorator `@secure()` to declare the parameter as secret. For some reason, the type system still sees this as a regular string. I wonder if this is related to #1535. 

![image](https://user-images.githubusercontent.com/11078275/108604859-3bf90a00-7365-11eb-8ca7-e88e2a34a76f.png)

One potential improvement (not sure how realistic it is). I created the secret as a top-level child resource. I'd like to be able to just pass the resource `secret` like so:

```bicep
// create secret
resource secret 'Microsoft.KeyVault/vaults/secrets@2018-02-14' = { ... }

module secretDeploy 'module-with-secret.bicep' = {
  name: 'foo'
  params: {
    secret: secret
  }
}
```

Or

```bicep
module secretDeploy 'module-with-secret.bicep' = {
  name: 'foo'
  params: {
    secret: secret.getSecret()
  }
}
```

miqm commented:
@alex-frankel This is still work in progress, I wanted to get some insight if my approach to implementing this is the right way to go.

miqm commented:
As for the improvement, I initially though about this. First option seems to me to be very hard to implement in bicep, but second one is possible, but perhaps `asSecret()` would be better.

However this will not work in green field deployments. When I run validate (so even before deployment begins), I receive following error (key vault existed before):
```
{"error":{"code":"KeyVaultParameterReferenceSecretRetrieveFailed","message":"The secret of KeyVault parameter 'mySecret1' cannot be retrieved. Http status code: 'NotFound'. Error message: 'A secret with (name/id) mySecret1 was not found in this key vault. If you recently deleted this secret you may be able to recover it using the correct recovery command. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125182'. Please see https://aka.ms/arm-keyvault for usage details."}}
```
Or for non-existing key-vault (that is going to be deployed):
```
{"error":{"code":"KeyVaultParameterReferenceNotFound","message":"The specified KeyVault '/subscriptions/[GUID]/resourceGroups/bicep-rg/providers/Microsoft.KeyVault/vaults/kv-bicep-2' could not be found. Please see https://aka.ms/arm-keyvault for usage details."}}
```

Perhaps this is something to be fixed/changed on the Deployment Manager level.

miqm commented:
Also, a question here - how to implement decompilation? Existing resource we define using 2 parameters: name and scope. In ARM we had only 1 - id. 

I think of 3 options: 
1. put id in name field - as decompilation is a best-effort, user will need to correct the usage.
2. try to use some string functions to parse value provided there.
3. add ability to declare existing resource by id only.

For now I think I will go with 1. and 2. can be a future enhancement.

TarunSunkaraneni commented on src/Bicep.Core/Emit/ExpressionEmitter.cs:
```diff
@@ -182,6 +182,47 @@ public void EmitObjectProperties(ObjectSyntax objectSyntax, ISet<string>? proper
+                var objectSymbol = context.SemanticModel.GetSymbolInfo(instanceFunctionCall.BaseExpression);
+                if (objectSymbol is ResourceSymbol resource && resource.Type is ResourceType resourceType &&
+                            resourceType.Body is NamedObjectType objectType &&
+                            string.Equals(objectType.Name, "Microsoft.KeyVault/vaults", StringComparison.OrdinalIgnoreCase))
```
Note to change this to `StringComparer.OrdinalIgnoreCase.Equals(` and extract `Microsoft.KeyVaylt/vaults` to a constant in your final PR

miqm replied:
extracting to constant - yes, but is there a difference between `string.equals(v1,v2, StringComparer.OrdinalIngnoreCase)` and `StringComparer.OrdinalIgnoreCase.Equals`?

miqm replied:
@TarunSunkaraneni  I simplified this a bit, Now i just check for the function name, not for the type on which it's being called.

miqm commented on src/Bicep.Core/Syntax/ParameterDeclarationSyntax.cs:
```diff
@@ -1,95 +1,104 @@
-// Copyright (c) Microsoft Corporation.
```
I didn't change anything it this file, I think when I was debugging in VS, this file opened and I hit to unify line endings.

miqm commented on src/Bicep.Core/TypeSystem/Az/AzResourceTypeProvider.cs:
```diff
@@ -125,12 +126,32 @@ private static ObjectType SetBicepResourceProperties(ObjectType objectType, Reso
+                functions);
+        }
+
+        private static IEnumerable<FunctionOverload> GetTypeBicepFunctions(string name)
```
This is the function that we can use to define more functions that can be called on particular types.

miqm commented on src/Bicep.Core/TypeSystem/NamedObjectType.cs:
```diff
@@ -9,11 +9,16 @@ namespace Bicep.Core.TypeSystem
+            
+        }
+
+        public NamedObjectType(string name, TypeSymbolValidationFlags validationFlags, IEnumerable<TypeProperty> properties, ITypeReference? additionalPropertiesType, TypePropertyFlags additionalPropertiesFlags, FunctionResolver functionResolver)
```
During building process, there are 2 places where we convert one NamedObjectType instance into a fresh one and the MethodResolver was being lost. To properly surface methods added on a type in AzResourceTypeProvider I needed to preserve it.

miqm commented on src/Bicep.Core/TypeSystem/TypeValidator.cs:
```diff
@@ -466,13 +478,13 @@ private static TypeSymbol NarrowObjectType(ITypeManager typeManager, ObjectSynta
             }
 
-            return new NamedObjectType(targetType.Name, targetType.ValidationFlags, narrowedProperties, targetType.AdditionalPropertiesType, targetType.AdditionalPropertiesFlags);
+            return new NamedObjectType(targetType.Name, targetType.ValidationFlags, narrowedProperties, targetType.AdditionalPropertiesType, targetType.AdditionalPropertiesFlags, targetType.MethodResolver);
```
preserving MethodResolver so type functions are not lost.

anthony-c-martin commented:
Hey @miqm - @alex-frankel, @majastrz, @shenglol and I had a chat about this PR yesterday. We thought that the approach we should follow is to use the `FunctionFlags` attribute and adding validation to `SymbolValidator` instead of adding a new type to the type system. Reasoning was:
1. This is more a question of **where** the function should be permitted than a question of types, and we have existing infrastructure to perform similar checks (for example, dealing with functions that are only allowed in parameter default values).
2. We're trying to be careful about expanding types, as they can add significant maintenance burden by expanding the type assignability matrix. We may want to add the notion of secure-ness in the future, but I think we would want it to be an attribute of a type, rather than a type itself.

Sorry for the delay in getting back to you - we really appreciate your contribution, and please feel free to check in with us for any pointers!

miqm commented:
Thanks for the hint, I'll try to rework my PR. What about the decompilation approach? Perhaps we can catch up on Teams sometime next week?

miqm commented:
After some time merging/rebasing with main (💘 LF) I managed to put some improvements.

After talking with @anthony-c-martin I renamed and changed the introduced type - it's now `keyVaultSecretReference` (was `secret` before) so not to mix up with secret or secure objects.

Function flags are used to indicate placement inside module params, although I needed to make TypeAssignmentVisitor capable of using them when checking instance functions. In the visitor at the point of validating function placement we do not have information about the type of the parameter and therefore we're unable to detect if we're using function in param that accepts it.

I didn't want to write new SyntaxVisitor and introduce a mechanism for validating function flags which might add even more complexity, have performance drawbacks and could be easily mixed up with existing one that is basically split between TypeAssignmentVisitor and NameBindingVisitor. 
Perhaps this is room for improvement and making function flags to be validated in a single place, but I think this PR is not the place for it.
Right now I added checking FunctionFlags to TypeAssignmentVisitor so it can be used with InstanceFunctions (previously Default was always used).

The decompilation is not done as it's blocked by #1722 - Suggestion here is to implement it later - after having nice way of referencing external resources by id only rather than name/scope pair (which here would require some extra complexity and smart detecting "what author had id mind" style.

Also, added some scenario tests to cover proper and wrong use cases, so I think we're ready for the proper review.

anthony-c-martin commented on src/Bicep.Core/Syntax/ParameterDeclarationSyntax.cs:
```diff
@@ -64,6 +65,8 @@ public TypeSymbol GetDeclaredType()
 
         public TypeSymbol GetAssignedType(ITypeManager typeManager, ArraySyntax? allowedSyntax)
         {
+            static bool IsSecure(SyntaxBase? value) => value is BooleanLiteralSyntax boolLiteral && boolLiteral.Value;
```
What about `IsBooleanTrue` instead of `IsSecure`?

miqm replied:
modified.

anthony-c-martin commented on src/Bicep.Core/TypeSystem/Az/AzResourceTypeProvider.cs:
```diff
@@ -125,12 +126,32 @@ private static ObjectType SetBicepResourceProperties(ObjectType objectType, Reso
+            {
+                case "microsoft.keyvault/vaults":
+                    yield return new FunctionOverloadBuilder("getSecret")
+                        .WithReturnType(LanguageConstants.KeyVaultSecretReference)
```
Why does this return a `KeyVaultSecretReference` type (extension of object) rather than a secure string with `AllowKeyVaultSecretReferenceAssignment` flag?

miqm replied:
The main reason behind this (and introducing a type) was reason, that I wasn't able to pin-point allowed location to exactly the parameter of type string with secure decorator. The closest I could get was to detect that we're in an object property which is params inside module.

Then, if I'd use only string with ValidationFlag only, I was getting errors like: `cannot assign type string to type string`. A
Of course I could add another cases and additional diagnostic errors to handle this error, but in my opinion this is even worse (branching new "special" cases in various, distinct places) than introducing type and taking advantage of the type checking mechanism.

miqm replied:
@anthony-c-martin @majastrz here is where we ensure that function is only available for using on keyVault type.

anthony-c-martin commented on src/Bicep.Core/TypeSystem/TypeValidator.cs:
```diff
@@ -85,6 +85,14 @@ public static bool AreTypesAssignable(TypeSymbol sourceType, TypeSymbol targetTy
                     // string literals can be assigned to strings
                     return targetType.Name == LanguageConstants.String.Name;
 
+                case PrimitiveType _ when targetType.Name == LanguageConstants.StringTypeName && sourceType is KeyVaultSecretReferenceType:
+                    // key vault reference type can be assigned only to string type parameters with secure decorator
+                    return targetType.ValidationFlags.HasFlag(TypeSymbolValidationFlags.AllowKeyVaultSecretReferenceAssignment);
+
+                case KeyVaultSecretReferenceType _:
+                    //key vault secret reference type is source type only.
+                    return false;
+
```
Couldn't we just use `LanguageConstants.SecureString` here by augmenting string assignment with the requirement that you can't assign secure to insecure? Or does that introduce other problems?

miqm replied:
It did introduce problem with ambiguous errors like 'cannot assign string to string' when using against non-secure parameter.

It was also being allowed inside string interpolation.

anthony-c-martin commented on src/Bicep.Core/TypeSystem/KeyVaultSecretReferenceType.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Bicep.Core.TypeSystem
+{
+    class KeyVaultSecretReferenceType : ObjectType
```
I feel like adding a new type needs **really** good justification as it adds significant complexity in the long run. Is there a reason you can't use the `SecureString` modified string type added in this PR for type checks?

miqm replied:
I was thinking about this (the additional flag as we spoke) but the amount of places I'd need to add handling this (to forbid from using it as regular string) to not allow using the function was adding even more complexity. And more important - the error thrown was saying about trying to assign string to string, which was odd.

majastrz replied:
I'm not convinced at all that enforcing that `getSecret` method only is allowed with secure parameters actually belongs in the core compiler logic. Feels like this should be the job of the linter to catch. (Of course, I would want that linter rule to default to error and be run by default.)

miqm replied:
I think it should, as we change what is being compiled. Instead a `value` object we emit `reference`. It’s not just a runtime value change, it’s changing compiler logic.

majastrz replied:
The transpiling of `value` into `reference` must be in the core compiler for sure. I'm talking about blocking usage of `getSecret()` with non-secure parameters. That part doesn't have to be in the core compiler and could be delegated to a linter rule.

anthony-c-martin reviewed (changes requested):
Just had some questions, mostly about the type system changes.

anthony-c-martin commented on src/Bicep.Core.IntegrationTests/ScenarioTests.cs:
```diff
@@ -957,5 +957,134 @@ public void Test_Issue1432()
         }
+
+        [TestMethod]
+        public void Test_Issue1028_1()
```
I think it would be worth breaking out these test methods into a dedicated new test class `KeyVaultSecretReferenceTests.cs` and making the method names a bit more descriptive - e.g. `KeyVault_secrets_can_be_passed_to_module_parameters()`, `KeyVault_secrets_cannot_be_exposed_as_module_outputs()` etc

miqm replied:
Yes, will do. I'll add a folder for scenario tests, as we'll be adding more and more separate files will reduce merge conflicts.

miqm replied:
moved to a file in folder called "Scenarios"

anthony-c-martin commented on src/Bicep.Core.IntegrationTests/ScenarioTests.cs:
```diff
@@ -957,5 +957,134 @@ public void Test_Issue1432()
+module secret 'secret.bicep' = {
+  name: 'secret'
+  params: {
+    mySecret: kv.getSecret('mySecret')
```
Would you mind adding an example of this syntax to the baselines at `/src/Bicep.Core.Samples/Files/Modules_CRLF` - just as we often use them to understand changes to the syntax structure, and I think you're the first user of object methods 😄

anthony-c-martin replied:
In fact, some negative tests under `/src/Bicep.Core.Samples/Files/InvalidModules_LF` would probably also be valuable, as they also help to visually understand what the diagnostics will look like.

miqm replied:
I added some negative tests and the warnings didn't pop up in all places where they should... will be fixing this.

miqm commented:
@anthony-c-martin I just wanted to add a bit more regarding introducing the type.

My main flow is that the `getSecret()` method does **not** return a `securestring`. Module param with type `securestring` (`string` with `@secure()` decorator) is capable of accepting a `StringLiteral` `string` (primitive) and thanks to the additional validation flag - `keyVaultSecretReference`. 

Even in the ARM output we have a different structure passed to the parameter: instead `{"value": "text"}` we're building `"{"reference":{...} }`. 

If the method would return a `string` or `securestring`, it might give a false impression, that we can assign it to a variable, output, pass to a parameter and what's more - that the method does actually gets a secret from key vault and we have it where we want it.

With this type, we're saying - hey - this is a reference to the key vault secret - you can use it in the module params. And there's no confusion that this is a string. Developer would think more what she or he can do with this type, rather than: "ok, this is string, so I can use it in interpolation, get count of it, concat it, etc.".

miqm commented:
I've converted it back to draft, type checking gives me more headache than I thought it does. Need to work more on this. Plus the issues with test suite.

It seemed quite straightforward at the beginning...

miqm commented:
Blocked by #1915

miqm commented on src/Bicep.Core.Samples/Files/InvalidOutputs_CRLF/Completions/decorators.json:
```diff
@@ -1,152 +1,19 @@
-[
```
Not sure why this is being removed...

anthony-c-martin replied:
I think because this decorator (and the `decoratorsPlusNamespace` example) was previously not followed by any declaration in this file, the list of suggestions was previously giving the full list of decorators that are valid on **any** declaration.

Because you added a resource declaration below the decorator [here](https://github.com/Azure/bicep/blob/7d720e7c317a8401bb0a4e9da151f1f3035d8d69/src/Bicep.Core.Samples/Files/InvalidOutputs_CRLF/main.bicep#L139), the suggestions are being filtered down to only decorators that are appropriate on a resource declaration.

miqm replied:
Ok, I see. Shall I move it upwards?

anthony-c-martin replied:
I think so, and probably just worth adding a comment like this one to explain why the decorators are at the end of the file so that it's clear to the next person editing this file: https://github.com/Azure/bicep/blob/187d4d2047dc83c69695ba79761f552bcb00c319/src/Bicep.Core.Samples/Files/InvalidExpressions_LF/main.bicep#L251

miqm replied:
Note here - deletion and removal is caused by CRLF file being check-in in the main branch, while my git did CRLF -> LF conversion.

miqm replied:
@anthony-c-martin - should I revert this to CRLF (and perhaps add .gitattribute to preserve line endings) or we're fine with json diagnostics are checked-in with LF?

miqm commented on src/Bicep.Core.Samples/Files/InvalidOutputs_CRLF/Completions/decoratorsPlusNamespace.json:
```diff
@@ -1,166 +1,33 @@
-[
```
Same here, I feel that this should remain as is...

miqm commented on src/Bicep.Core.Samples/Files/Modules_CRLF/main.formatted.bicep:
```diff
@@ -253,3 +253,17 @@ module propertyLoopInsideParameterValueInsideModuleLoop 'modulea.bicep' = [for t
+
+// Key Vault Secret Reference
+
+resource kv 'Microsoft.KeyVault/vaults@2019-09-01' existing = {
```
Would be good to add example of key vault from another RG

miqm commented:
@anthony-c-martin Could you check why build is failing? It seems like SetBaseline script somehow does not detect changes made.

anthony-c-martin commented:
> @anthony-c-martin Could you check why build is failing? ~It seems like SetBaseline script somehow does not detect changes made.~
> 
> I did some debugging and it's not problem with SetBaseline but with following [code](https://github.com/azure/bicep/blob/main/src/Bicep.LangServer.IntegrationTests/ReferencesTests.cs#L59):

I'm pretty sure the difference in behavior you're seeing is because the test framework is using `TestResourceTypeProvider`, whereas the extension is using `AzResourceTypeProvider`. The former doesn't do any sort of caching, whereas the latter does.

To test this out, I modified your branch to change the `TestResourceTypeProvider` to add a basic cache, and the failing test passed:
![image](https://user-images.githubusercontent.com/38542602/114344846-937a5180-9b2e-11eb-8e84-8e04eabe2a99.png)

It's starting to feel like we need to better commonize the logic between the test `IResourceTypeProvider` implementations and the real implementations, or we could miss some subtle bugs...

miqm commented:
So what's the next step? Add cache or commonize logic or use AzResourceTypeProvider in tests like here https://github.com/Azure/bicep/pull/2207/files#diff-da4592c33591bf2711db9fad31b8638241d159f5d2bbd6afdf10f1d61e0e397fR120 ?

anthony-c-martin commented:
> So what's the next step? Add cache or commonize logic or use AzResourceTypeProvider in tests like here https://github.com/Azure/bicep/pull/2207/files#diff-da4592c33591bf2711db9fad31b8638241d159f5d2bbd6afdf10f1d61e0e397fR120 ?

I've created #2222, which I think should address this. I'll test it against your branch before merging.

miqm commented:
Merged main into branch, tests are passing 🎉  Waiting for further comments on the code.

anthony-c-martin commented on docs/spec/modules.md:
````diff
@@ -71,4 +71,44 @@ module publicIp './publicIpAddress.bicep' = {
+```
+
+### Notes
+* Key Vault must have `enabledForDeployment` property set to `true`
````
Are there any docs we could link to on this?

miqm replied:
https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/template-tutorial-use-key-vault#prepare-a-key-vault

alex-frankel replied:
https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/key-vault-parameter?tabs=azure-cli

anthony-c-martin commented on src/Bicep.Core.IntegrationTests/Scenarios/ParamKeyVaultSecretReferenceTests.cs:
```diff
@@ -0,0 +1,218 @@
+output exposed string = kv.getSecret('mySecret','secretversionguid')
+"));
+
+            diags.Should().NotBeEmpty();
```
Mind adding a specific assertion for the diagnostics you're expecting (here and in other places)? It helps avoid accidental regressions in the error handling. See here for an example: https://github.com/Azure/bicep/blob/d3ea3e822d128d29b28101b54086c67bbfb7a10b/src/Bicep.Core.IntegrationTests/ScenarioTests.cs#L1237-L1244

anthony-c-martin commented on src/Bicep.Core/Emit/ExpressionEmitter.cs:
```diff
@@ -314,6 +314,46 @@ public void EmitObjectProperties(ObjectSyntax objectSyntax, ISet<string>? proper
+
+                    return;
+                }
+                throw new InvalidOperationException("Cannot emit parameter's KeyVault secret reference.");
```
Presumably this'll never actually happen in practice (e.g. you've already performed validation and blocked emitting). Mind confirming?

miqm replied:
Yes, it could be thrown only when function was called on something else than a Resource.

anthony-c-martin commented on src/Bicep.Core/Syntax/ObjectSyntaxExtensions.cs:
```diff
@@ -115,5 +115,10 @@ public static ObjectSyntax DeepMerge(this ObjectSyntax? sourceObject, ObjectSynt
                     : mergedObject);
         }
+
+        public static DecoratorSyntax? SafeGetDecoaratorByName(this StatementSyntax syntax, string name)
```
[nit] typo

anthony-c-martin commented on src/Bicep.Core/Syntax/ObjectSyntaxExtensions.cs:
```diff
@@ -115,5 +115,10 @@ public static ObjectSyntax DeepMerge(this ObjectSyntax? sourceObject, ObjectSynt
+
+        public static DecoratorSyntax? SafeGetDecoaratorByName(this StatementSyntax syntax, string name)
+        {
+            return syntax.Decorators.FirstOrDefault(d => d.Expression is FunctionCallSyntax func && string.Equals(func.Name.IdentifierName, name, System.StringComparison.OrdinalIgnoreCase));
```
This won't work for namespaced decorators - e.g.:
```bicep
@sys.secure()
```

shenglol replied:
Yeah, it's also possible that in the future some other namespace add a `secure` decorator, so we may need to get the decorator function symbol first and search in the `sys` namespace:
```C#
return syntax.Decorators.FirstOrDefault(decoratorSyntax =>
{
    if (this.binder.GetSymbolInfo(decoratorSyntax.Expression) is FunctionSymbol functionSymbol)
    {
        var argumentTypes = this.GetRecoveredArgumentTypes(decoratorSyntax.Arguments).ToArray();
        var decorator = this.binder.FileSymbol.ImportedNamespaces["sys"].Type.DecoratorResolver
            .GetMatches(functionSymbol, argumentTypes)
            .SingleOrDefault();

        if (decorator?.Overload.Name == name)
        {
            return true;
        }
    }

    return false;
});
```

miqm replied:
Ok, I see the issue. I'd need to move checking decorator/object modifier from ParameterDeclarationSyntax up to TypeAssignmentVisitor and alter the type/flag there if certain decorator/modifier is found.

TBH, similar to getSecret function we should make decorator builder to define (via lambda) how the type is going to change if the decorator it's being used, rather than hard-coding type logic in the TypeAssignmentVisitor.

miqm replied:
@shenglol - the code you provided does not work.
`this.binder.GetSymbolInfo(decoratorSyntax.Expression)` is returning null.

I as you use same code to check for allowed decorator I checked with the 'stock' version and I get no error when I provide a non-allowed value when using `@sys.allowed()` decorator.

anthony-c-martin commented on src/Bicep.Core/TypeSystem/Az/AzResourceTypeProvider.cs:
```diff
@@ -197,12 +198,31 @@ private static ObjectType SetBicepResourceProperties(ObjectType objectType, Reso
+                case "microsoft.keyvault/vaults":
+                    yield return new FunctionOverloadBuilder("getSecret")
+                        .WithReturnType(LanguageConstants.KeyVaultSecretReference)
+                        .WithDescription("References a secret from this key vault to be binded to a secure string module parameter")
```
What about "Gets a reference to a key vault secret, which can be provided to a secure string module parameter"?

alex-frankel replied:
+1

anthony-c-martin commented on src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs:
```diff
@@ -554,20 +568,32 @@ public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
                     types.Add(keyType);
                 }
 
+                var vistingModuleParams = visitingModule
+                    && !allowedFunctionFlags.HasFlag(FunctionFlags.ModuleParamsAssignmentOnly)
+                    && syntax.Key is IdentifierSyntax identifierSyntax
+                    && string.Equals(identifierSyntax.IdentifierName, "params", StringComparison.OrdinalIgnoreCase);
+                if (vistingModuleParams)
+                {
+                    allowedFunctionFlags |= FunctionFlags.ModuleParamsAssignmentOnly;
+                }
                 var valueType = typeManager.GetTypeInfo(syntax.Value);
+                if (vistingModuleParams)
+                {
+                    allowedFunctionFlags &= ~FunctionFlags.ModuleParamsAssignmentOnly;
+                }
```
This logic doesn't feel strictly related to type checking, and introducing new state into the type checker feels risky as the recursion is already very complicated. Would it be possible to implement this in the `EmitLimitationCalculator` class? Or is there anywhere else that this could fit?

miqm replied:
I remember struggling to place it elsewhere but a the end typeVisitor was the only place I had access to the information which function overload was actually used.

the best would be to put it in nameBindingVisitor where we check other flags, but that visitor does not visit instance Function calls. For instance functions, flags are checked here. Currently default is hardcoded, so there’s no flag checking on instance functions at all.

In fact we do not need this at all. Type assignment checking will do the work, as we can assign function result only to a string param with secure decorator. I initially planned to do so, but I added this upon review request.

majastrz replied:
This check would only be needed because we have no way to emit other usages, so perfect fit for the EmitLimitation* classes (if we end up needing it after other conversations).

miqm replied:
Thing is that this fragment is not related to emitting. It’s for function flags which are used to limit functions available when assigning function result to a type.

miqm replied:
Function flags mechanism here, in TypeAssignmentVisitor could be useful in future if we’d like to have list* functions as instance functions, as they cannot be used in all places.

miqm replied:
As discussed, moved entire logic to a separate visitor

majastrz commented on docs/spec/modules.md:
```diff
@@ -71,4 +71,44 @@ module publicIp './publicIpAddress.bicep' = {
+### Notes
+* Key Vault must have `enabledForDeployment` property set to `true`
+* Key Vault and secret must exist before entire deployment starts.
+* Secret version is optional
```
>optional [](http://example.com/codeflow?start=20&length=8)

should mention that it defaults to latest version if omitted.

majastrz commented on src/Bicep.Core/LanguageConstants.cs:
```diff
@@ -123,6 +125,7 @@ public static TypeSymbol CreateParameterModifierType(TypeSymbol primitiveType, T
             return new ObjectType($"ParameterModifier<{allowedValuesType.Name}>", TypeSymbolValidationFlags.Default, CreateParameterModifierProperties(primitiveType, allowedValuesType), additionalPropertiesType: null);
         }
 
+        public const string ParameterModifierSecureName = "secure";
```
>ParameterModifierSecureName [](http://example.com/codeflow?start=28&length=27)

nit - can you move it up and group it with the other constants?

majastrz commented on src/Bicep.Core/Emit/ExpressionEmitter.cs:
```diff
@@ -314,6 +314,46 @@ public void EmitObjectProperties(ObjectSyntax objectSyntax, ISet<string>? proper
 
+        public void EmitModuleParameterValue(SyntaxBase syntax)
+        {
+            if (syntax is InstanceFunctionCallSyntax instanceFunctionCall && string.Equals(instanceFunctionCall.Name.IdentifierName, "getSecret", StringComparison.Ordinal))
```
>StringComparison.Ordinal [](http://example.com/codeflow?start=146&length=24)

Should use `LanguageConstants.IdentifierComparison`. It's the same functionally but it makes it easy to find all the places where we compare identifiers.

majastrz commented on src/Bicep.Core/Emit/ExpressionEmitter.cs:
```diff
@@ -314,6 +314,46 @@ public void EmitObjectProperties(ObjectSyntax objectSyntax, ISet<string>? proper
 
+        public void EmitModuleParameterValue(SyntaxBase syntax)
+        {
+            if (syntax is InstanceFunctionCallSyntax instanceFunctionCall && string.Equals(instanceFunctionCall.Name.IdentifierName, "getSecret", StringComparison.Ordinal))
```
>string.Equals(instanceFunctionCall.Name.IdentifierName, "getSecret" [](http://example.com/codeflow?start=77&length=67)

@anthony-c-martin Don't we have a helper somewhere that checks the bound symbol to verify that the function is the function we think it is?

majastrz replied:
My main concern is that in the future we could have a `getSecret` function in one of the imported namespaces that behaves differently to the method on a KV.

anthony-c-martin replied:
I can't think of anything immediately. The next best thing would be to type check the base expression on the InstanceFunctionCallSyntax and verify it is a resource type, and also that it's the particular resource type you expect (KV).

miqm replied:
This is being checked. You function getSecret is only available on keyVault type.

miqm replied:
I’d need to add to test for it.

anthony-c-martin replied:
> This is being checked. Function getSecret is only available on keyVault type.

It is today, but I think Marcin's concern is that if we're not specific now, we're going to potentially generate invalid code if we ever add a function named `getSecret` to any other object type.

miqm replied:
true, I'll add a check then. I had it before AFAIR,  but got lost when refactoring :)

miqm replied:
I’ve an idea how to improve it.

what if instead writing emit instructions here I’d modify code to call a lambda that is associated with the type that is going to be assigned to parameter? I did something similar in includeContent function (still in progress though). The lambda would be defined with the function builder. WDYT?

alex-frankel commented on src/Bicep.Core/Diagnostics/DiagnosticBuilder.cs:
```diff
@@ -997,6 +997,16 @@ public ErrorDiagnostic VariableLoopsRuntimeDependencyNotAllowed(IEnumerable<stri
+            public ErrorDiagnostic FunctionOnlyValidInModuleParameterAssignment(string functionName) => new(
+                TextSpan,
+                "BCP176",
+                $"Function \"{functionName}\" is not valid at this location. It can only be used in assigning value to a module parameter.");
```
nit - Can we tweak the second sentence:

"It can only be used when assigning a value to a module parameter"

alex-frankel commented on src/Bicep.Core/Diagnostics/DiagnosticBuilder.cs:
```diff
@@ -997,6 +997,16 @@ public ErrorDiagnostic VariableLoopsRuntimeDependencyNotAllowed(IEnumerable<stri
+            public ErrorDiagnostic TypeNotValidInStringInterpolation(TypeSymbol type) => new(
+                TextSpan,
+                "BCP177",
+                $"Type \"{type.Name}\" cannot be used inside String interpolation.");
```
I think the 's' in "string interpolation" should be lower cased

miqm commented on src/Bicep.Core/Emit/ExpressionEmitter.cs:
```diff
@@ -314,6 +314,46 @@ public void EmitObjectProperties(ObjectSyntax objectSyntax, ISet<string>? proper
+            if (syntax is InstanceFunctionCallSyntax instanceFunctionCall && string.Equals(instanceFunctionCall.Name.IdentifierName, "getSecret", StringComparison.Ordinal))
+            {
+                var objectSymbol = context.SemanticModel.GetSymbolInfo(instanceFunctionCall.BaseExpression);
+                if (objectSymbol is ResourceSymbol resource)
```
check if this is key vault

miqm commented on src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs:
```diff
@@ -39,13 +39,15 @@ public TypeAssignment(ITypeReference reference, IEnumerable<Diagnostic> diagnost
         private readonly TypeManager typeManager;
         private readonly IBinder binder;
         private readonly IDictionary<SyntaxBase, TypeAssignment> assignedTypes;
+        private readonly IDictionary<FunctionSymbol, FunctionOverload> matchedFunctions;
```
@majastrz This is new - an attempt to get which function was used to resolve the type

majastrz replied:
Do we even need to cache these? I think the highest number of overloads we have for a function is like 4, so computing it every time we need it shouldn't be too bad.

Because of the `TryAdd()` comment below, lookups might fail against this dictionary, so what's the point?

miqm replied:
I didn't want do duplicate code responsible for matching function overload into the different visitor and avoid case where potentially different functions are matched.

miqm replied:
not valid anymore, functionality removed.

miqm commented on src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs:
```diff
@@ -1179,8 +1208,11 @@ private static void CollectErrors(List<ErrorDiagnostic> errors, ITypeReference r
-                return matches.Single().ReturnTypeBuilder(argumentSyntaxes);
+                var matchedOverload = matches.Single();
+                // record which function was used to get the type                
+                matchedFunctions.TryAdd(function, matchedOverload);
```
I used here TryAdd - I was oddly getting duplicate inserts for this same symbol. Is this something that could happen? I thought one symbol can return only one type, and type assignment caching should prevent second call...

majastrz replied:
I think some of the function symbols are created on a demand every time.

miqm replied:
ok, I need to dig a bit deeper then or perhaps change dictionary to use syntax as a key.

miqm replied:
@majastrz - I removed this part and exposed PlacementFlags through FunctionSymbol, similar to FunctionFlags

miqm commented on src/Bicep.Core/TypeSystem/TypeKind.cs:
```diff
@@ -63,5 +63,10 @@ public enum TypeKind
+        /// <summary>
+        /// Reference to KeyVault secret
+        /// </summary>
+        KeyVaultSecretReference
```
To delete.

miqm commented:
@majastrz @alex-frankel - I think this is ready now for final review, I implemented all we talked about and added tests with module loops as suggested.

Only thing that is not working is support for `@sys.secure()` decorator (with a namespace), but this does not work for `@sys.allowed()` as well, so I'll create a separate issue to cover that case (I used this same code as per @shenglol suggestion).

majastrz commented on src/Bicep.Core/Emit/FunctionPlacementValidatorVisitor.cs:
```diff
@@ -0,0 +1,146 @@
+            {
+                var functionOverload = semanticModel.TypeManager.GetMatchedFunctionOverload(functionSymbol);
+                if (functionOverload is not null && functionOverload.PlacementFlags.HasFlag(FunctionPlacementFlags.ModuleSecureParameterOnly))
+                { // we can check placement only for funtions that were matched and has a proper placement flag
```
>// [](http://example.com/codeflow?start=18&length=2)

nit - indentation

miqm replied:
Fixed.

majastrz commented on src/Bicep.Core/Emit/FunctionPlacementValidatorVisitor.cs:
```diff
@@ -0,0 +1,146 @@
+            var vistingModuleParams = visitedElementsStack.Contains(VisitedElement.Module)
+                && !(visitedElementsStack.TryPeek(out var head) && head == VisitedElement.ModuleParams)
+                && syntax.Key is IdentifierSyntax identifierSyntax
+                && string.Equals(identifierSyntax.IdentifierName, "params", StringComparison.OrdinalIgnoreCase);
```
>StringComparison.OrdinalIgnoreCase [](http://example.com/codeflow?start=76&length=34)

Should use the identifier comparer.

miqm replied:
I think types are case-insensitive - identifier comparer is case-sensitive. perhaps I should add typeComparer alongside identifierComparer?

miqm replied:
Fixed.

majastrz commented on src/Bicep.Core/Emit/FunctionPlacementValidatorVisitor.cs:
```diff
@@ -0,0 +1,146 @@
+            var vistingModuleParams = visitedElementsStack.Contains(VisitedElement.Module)
+                && !(visitedElementsStack.TryPeek(out var head) && head == VisitedElement.ModuleParams)
+                && syntax.Key is IdentifierSyntax identifierSyntax
+                && string.Equals(identifierSyntax.IdentifierName, "params", StringComparison.OrdinalIgnoreCase);
```
>"params" [](http://example.com/codeflow?start=66&length=8)

worth adding a constant to match the name of this with the same property where we calculate the module type?

miqm replied:
It turned out constant was already there, I used it.

majastrz commented on src/Bicep.Core/Emit/FunctionPlacementValidatorVisitor.cs:
```diff
@@ -0,0 +1,146 @@
+        private readonly Stack<(SyntaxBase syntax, Symbol? symbol)> visitedSymbolsStack = new();
+        private readonly Stack<VisitedElement> visitedElementsStack = new();
+
+        private class VisitorSymbolScope : IDisposable
```
>IDisposable [](http://example.com/codeflow?start=43&length=11)

This is a really neat way of managing the stack :)

miqm replied:
thanks :)

miqm replied:
@majastrz - I extracted it into a separate class so it can be reused in other visitors if needed.

majastrz commented on src/Bicep.Core/TypeSystem/FunctionPlacementFlags.cs:
```diff
@@ -0,0 +1,23 @@
+    /// Flags that may be placed on functions to indicate where they can be placed in semantic tree
+    /// </summary>
+    [Flags]
+    public enum FunctionPlacementFlags
```
>FunctionPlacementFlags [](http://example.com/codeflow?start=16&length=22)

`FunctionFlags` already has a bunch of placement-like flags. Do we need the second enum?

miqm replied:
Function flags are used in type assignment. adding a flag there and assigning it to function overload will trigger the mechanism and prevent using it - type assignment visitor will want function with default flag no with something different.

miqm replied:
@majastrz - is this explanation sufficient?

majastrz replied:
Both `FunctionFlags` and `FunctionPlacementFlags` are declared and assigned together in every place they appear, so their lifecycles appear to be identical.

If you look at these values of `FunctionFlags`, all of them control placement of the functions:
- `ParamDefaultsOnly`
- `ParameterDecorator`
- `VariableDecorator`
- `ResourceDecorator`
- `ModuleDecorator`
- `OutputDecorator`
- `ResourceOrModuleDecorator`
- `AnyDecorator`

In fact it seem like the majority of the enum values are about placement, so it's kind of weird to have a second set of flags also about placement.

You mentioned that adding the flag to `FunctionFlag` would trigger the type checking mechanism? Can you point me to the code that's doing it? Maybe we can tweak that to avoid it?

miqm replied:
it's here: https://github.com/Azure/bicep/blob/de8fd8ed4d4aefd6519fe726953f8f255669e1ae/src/Bicep.Core/Semantics/SymbolValidator.cs#L105

In fact I can see now that additional flag will not impact the binding mechanism, but we will end up in having flags checked in 2 places.

miqm replied:
@majastrz - as discussed, I merged it into one enum (Function Flags).

majastrz commented on src/Bicep.Core/TypeSystem/Az/AzResourceTypeProvider.cs:
```diff
@@ -225,12 +226,31 @@ private static ObjectType SetBicepResourceProperties(ObjectType objectType, Reso
+                functions);
+        }
+
+        private static IEnumerable<FunctionOverload> GetTypeBicepFunctions(ResourceTypeReference resourceType)
```
>GetTypeBicepFunctions [](http://example.com/codeflow?start=53&length=21)

Can you rename to `GetBuiltInMethods` or `GetBicepMethods`?

miqm replied:
Renamed to GetBicepMethods

majastrz commented on src/Bicep.Core.Samples/Files/InvalidModules_LF/main.diagnostics.bicep:
```diff
@@ -545,7 +545,7 @@ module wrongModuleParameterInLoop2 'modulea.bicep' = [for (x,i) in emptyArray:{
     stringParamA: 'test'
     stringParamB: 'test'
     notAThing: 'test'
-//@[4:13) [BCP037 (Error)] No other properties are allowed on objects of type "params". |notAThing|
```
>allowed [](http://example.com/codeflow?start=51&length=7)

We might be losing coverage on the original error here. Could you undo the change to modulea and add a similar moduleb for your test instead?

miqm replied:
Fixed.

majastrz commented on src/Bicep.Core.Samples/Files/InvalidVariables_LF/main.diagnostics.bicep:
```diff
@@ -321,3 +321,12 @@ var loopExpression = union([for thing in stuff: 4], [for thing in stuff: true])
+  name: 'testkeyvault'
+}
+
+var keyVaultSecretVar = kv.getSecret('mySecret')
```
>getSecret [](http://example.com/codeflow?start=27&length=9)

Can you add a case where the function call is in deeper in an expression and also something involved complex objects?

miqm replied:
Done in arrays and object

majastrz commented on src/Bicep.Core.Samples/Files/InvalidOutputs_CRLF/main.diagnostics.bicep:
```diff
@@ -221,10 +221,22 @@ output noInnerLoopsInOutputs2 object = {
+  name: 'testkeyvault'
+}
+
+output keyVaultOutput string = kv.getSecret('mySecret')
```
>getSecret [](http://example.com/codeflow?start=34&length=9)

Can you add a case where the function call is in deeper in an expression and also something involved complex objects?

miqm replied:
Fixed.

majastrz reviewed (approved):
This is amazing! Thanks for getting this done!

v-mosh21 commented:
@miqm May I ask if there is any solution to solve the problem about using getSecret() to retrieve a secret value from a KeyVault provisioned in the same deployment?

I saw you have encountered a such problem previously as below:

For non-existing key-vault (that is going to be deployed):
{"error":{"code":"KeyVaultParameterReferenceNotFound","message":"The specified KeyVault '/subscriptions/[GUID]/resourceGroups/bicep-rg/providers/Microsoft.KeyVault/vaults/kv-bicep-2' could not be found. Please see https://aka.ms/arm-keyvault for usage details."}}

I found an other discussion on the same issue in this thread: https://githubmemory.com/repo/Azure/bicep/issues/4081.

Any suggestions will be very appreciated!
